### PR TITLE
feat(dup_enhancement#9): master replica support `on_query_last_checkpoint_info` rpc for follower replica

### PR DESCRIPTION
**Pull request by Shuo-Jia**
# Ref-Issue
https://github.com/apache/incubator-pegasus/issues/892

# Change
When follower start copy checkpoint, it need get checkpoint info firstly, so here add the rpc api support it

### CODE REVIEW

File: src/replica/replica_chkpt.cpp [hunk 1]
```diff
@@ -191,46 +198,29 @@ void replica::init_checkpoint(bool is_emergency)
         _stub->_counter_recent_trigger_emergency_checkpoint_count->increment();
 }
 
-// @ secondary
-void replica::on_copy_checkpoint(const replica_configuration &request,
```
Comment (hycdong) [inline src/replica/replica_chkpt.cpp:195]: I notice that you update the error code of this function, will it affect previous learn process?

Comment (Shuo-Jia) [reply src/replica/replica_chkpt.cpp:195]: it's not used in the past

File: src/replica/replica_stub.h [hunk 2]
```diff
@@ -135,7 +135,7 @@ class replica_stub : public serverlet<replica_stub>, public ref_counter
     void on_add_learner(const group_check_request &request);
     void on_remove(const replica_configuration &request);
     void on_group_check(group_check_rpc rpc);
-    void on_copy_checkpoint(copy_checkpoint_rpc rpc);
+    void on_query_last_checkpoint(query_last_checkpoint_info_rpc rpc);
```
Comment (hycdong) [inline src/replica/replica_stub.h:138]: Please add some comments to should it is used for duplication, same as the function definition in `replica.h`.

File: src/replica/replica_chkpt.cpp [hunk 3]
```diff
@@ -191,46 +198,29 @@ void replica::init_checkpoint(bool is_emergency)
...
-    }
-
     if (_app->last_durable_decree() == 0) {
-        response.err = ERR_OBJECT_NOT_FOUND;
+        response.err = ERR_PATH_NOT_FOUND;
         return;
     }
 
     blob placeholder;
     int err = _app->get_checkpoint(0, placeholder, response.state);
```
Comment (hycdong) [inline src/replica/replica_chkpt.cpp:212]: I suggest you can refactor function `get_checkpoint` in your further pull request, it seems only the last parameter useful, you can delete the first two parameters of this function.

File: src/replica/replica_chkpt.cpp [hunk 4]
```diff
@@ -191,46 +198,29 @@ void replica::init_checkpoint(bool is_emergency)
...
-        for (std::string &file_name : response.state.files) {
-            std::size_t last_splitter = file_name.find_last_of("/\\");
-            if (last_splitter != std::string::npos)
-                file_name = file_name.substr(last_splitter + 1);
-        }
+        response.base_local_dir = utils::filesystem::path_combine(
+            _app->data_dir(), checkpoint_folder(response.state.to_decree_included));
         response.address = _stub->_primary_address;
+        for (auto &file : response.state.files) {
+            file = file.substr(response.base_local_dir.length() + 1);
```
Comment (hycdong) [inline src/replica/replica_chkpt.cpp:222]: I am a little bit confused about the usage of `response.base_local_dir` and this `substr` sentence for `response.state.files`. Could you please explain it to me and add comments in code?

File: src/replica/replica_chkpt.cpp (hunk 3 above)
Comment (Shuo-Jia) [reply src/replica/replica_chkpt.cpp:212]: get_checkpoint is `pegasus` implementation

Comment (hycdong) [reply src/replica/replica_chkpt.cpp:212]: Yes, I have read the code of `get_checkpoint`, the first two parameters of this function is useless, you can refactor it in future.

File: src/replica/replica_chkpt.cpp (hunk 4 above)
Comment (Shuo-Jia) [reply src/replica/replica_chkpt.cpp:222]: it's same as `learn` code: 
https://github.com/XiaoMi/rdsn/blob/master/src/replica/replica_learn.cpp#L543

The peer need the `file path` list to copy

Comment (Shuo-Jia) [reply src/replica/replica_chkpt.cpp:225]: comments done.

File: src/replica/replica_chkpt.cpp [hunk 5]
```diff
@@ -191,46 +198,32 @@ void replica::init_checkpoint(bool is_emergency)
...
-        for (std::string &file_name : response.state.files) {
-            std::size_t last_splitter = file_name.find_last_of("/\\");
-            if (last_splitter != std::string::npos)
-                file_name = file_name.substr(last_splitter + 1);
-        }
+        // for example: base_local_dir = "./data" + "checkpoint.1024" = "./data/checkpoint.1024"
+        response.base_local_dir = utils::filesystem::path_combine(
+            _app->data_dir(), checkpoint_folder(response.state.to_decree_included));
         response.address = _stub->_primary_address;
+        for (auto &file : response.state.files) {
```
Comment (levy5307) [inline src/replica/replica_chkpt.cpp:222]: `for (const auto &file : response.state.files)`

**Review by levy5307 — APPROVED** [review]

**Pull request merged**